### PR TITLE
Gossipsub RPC Inspector tracker cache

### PR DESCRIPTION
**Pull request by kc1116**
This PR adds a cluster prefixed received tracker that will track the number of cluster prefix control messages received by a peer. The tracker uses herocache to store counter information for each peer. It also stores the active cluster IDs for cluster prefix topic validation.

This PR reduces the size of https://github.com/onflow/flow-go/pull/4189

### CODE REVIEW

**Review by yhassanzadeh13 — APPROVED** [review]

**Review by jordanschalm — COMMENTED** [review]
First round of comments

File: network/p2p/inspector/internal/cache/record.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import (
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// ClusterPrefixTopicsReceivedRecord cache record that keeps track of the amount of cluster prefixed
+// topics received from a peer.
+type ClusterPrefixTopicsReceivedRecord struct {
+	Identifier flow.Identifier
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/record.go]: https://github.com/onflow/flow-go/pull/4365/commits/189ef35883bd73c9cf264f2ce9fb44eb35a2cee5

File: network/p2p/inspector/internal/cache/record.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package cache
+
+import (
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// ClusterPrefixTopicsReceivedRecord cache record that keeps track of the amount of cluster prefixed
+// topics received from a peer.
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/record.go]: https://github.com/onflow/flow-go/pull/4365/commits/bb288c267d6d826c2945c8465f034fb189cfa6b6

File: network/p2p/inspector/internal/cache/active_cluster_ids_entity.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
+package cache
+
+import (
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// ActiveClusterIdsEntity is an entity that represents the active cluster IDs. This entity is used to leverage
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/active_cluster_ids_entity.go]: https://github.com/onflow/flow-go/pull/4365/commits/26c82949bd130aefd7e7ce25ed6625fdfb25e8ae

File: network/p2p/inspector/internal/cache/cache.go [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+// to keep track of the amount of cluster prefixed control messages received by a peer.
+type RecordCache struct {
+	// recordEntityFactory is a factory function that creates a new *RecordEntity.
+	recordEntityFactory recordEntityFactory
+	// c is the underlying cache.
+	c *stdmap.Backend
+	// decayFunc decay func used by the cache to perform decay on counters.
+	decayFunc preProcessingFunc
+	// activeClusterIdsCacheId identifier used to store the active cluster Ids.
+	activeClusterIdsCacheId flow.Identifier
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/cache.go]: https://github.com/onflow/flow-go/pull/4365/commits/26c82949bd130aefd7e7ce25ed6625fdfb25e8ae

File: network/p2p/inspector/internal/cache/cache.go [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+type preProcessingFunc func(recordEntity RecordEntity) (RecordEntity, error)
+
+// defaultDecayFunction is the default decay function that is used to decay the cluster prefixed topic received counter of a peer.
+func defaultDecayFunction(decay float64) preProcessingFunc {
+	return func(recordEntity RecordEntity) (RecordEntity, error) {
+		if recordEntity.Counter.Load() == 0 {
+			return recordEntity, nil
+		}
+
+		decayedVal, err := scoring.GeometricDecay(recordEntity.Counter.Load(), decay, recordEntity.lastUpdated)
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/cache.go]: https://github.com/onflow/flow-go/pull/4365/commits/5c6395d062526a6c3823785a25045865ef9465de

File: network/p2p/inspector/internal/cache/cache_entity.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	"time"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// RecordEntity is an entity that represents a tracking record that keeps track
+// of the amount of cluster prefixed topics received from a peer. This struct
+// implements the flow.Entity interface and uses a flow.Identifier created from
+// the records peer field for deduplication.
+type RecordEntity struct {
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/cache_entity.go]: https://github.com/onflow/flow-go/pull/4365/commits/1ff64af9a2ff8978e1b68527708d11c435911ee3

File: network/p2p/inspector/internal/cache/cache.go [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+}
+
+// Get returns the current number of cluster prefixed topcis received from a peer.
+// The record is initialized before the count is returned.
+// Before the count is returned it is decayed using the configured decay function.
+// Returns the record and true if the record exists, nil and false otherwise.
+// Args:
+// - originId: the origin id the sender of the control message.
+// Returns:
+// - The number of cluster prefix topics received after the decay and true if the record exists, 0 and false otherwise.
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/cache.go]: https://github.com/onflow/flow-go/pull/4365/commits/5c1ab564252c1e9e6e3a3b4236e88f71f20a0102

File: network/p2p/inspector/internal/cache/cache.go [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+// the adjustment func is applied that will increment the Counter.
+// It returns an error if the adjustFunc returns an error or if the record does not exist.
+// Assuming that adjust is always called when the record exists, the error is irrecoverable and indicates a bug.
+// Args:
+// - originId: the origin id the sender of the control message.
+// - adjustFunc: the function that adjusts the record.
+// Returns:
+//   - The number of cluster prefix topics received after the adjustment.
+//   - error if the adjustFunc returns an error or if the record does not exist (ErrRecordNotFound).
+//     All errors should be treated as an irrecoverable error and indicates a bug.
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/cache.go]: https://github.com/onflow/flow-go/pull/4365/commits/05f5cd4e5c44e0e0515bb92c9b7588d8ac9624c7

File: network/p2p/inspector/internal/cache/record.go [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+type ClusterPrefixTopicsReceivedRecord struct {
+	Identifier flow.Identifier
+	Counter    *atomic.Float64
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/record.go]: https://github.com/onflow/flow-go/pull/4365/commits/d0b0c1e69b99a98306f94831c84d7a8e0bb0d8fd

File: network/p2p/inspector/internal/cache/cache.go [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	herocache "github.com/onflow/flow-go/module/mempool/herocache/backdata"
+	"github.com/onflow/flow-go/module/mempool/herocache/backdata/heropool"
+	"github.com/onflow/flow-go/module/mempool/stdmap"
+	"github.com/onflow/flow-go/network/p2p/scoring"
+)
+
+var ErrRecordNotFound = fmt.Errorf("record not found")
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/cache.go]: https://github.com/onflow/flow-go/pull/4365/commits/738b86d9899f151e07b3f60217bdbb8ad7190fca

File: network/p2p/inspector/internal/cache/cache.go [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+
+func (r *RecordCache) decayAdjustment(entity flow.Entity) flow.Entity {
+	record, ok := entity.(RecordEntity)
+	if !ok {
+		// sanity check
+		// This should never happen, because the cache only contains RecordEntity entities.
+		panic(fmt.Sprintf("invalid entity type, expected RecordEntity type, got: %T", entity))
+	}
+	var err error
+	record, err = r.decayFunc(record)
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/cache.go:200]: https://github.com/onflow/flow-go/pull/4365/commits/41155e7d07ad88f457ae7c98dbdab19b1899ef75

File: network/p2p/inspector/internal/cache/cache.go [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+		decayedVal, err := scoring.GeometricDecay(recordEntity.Counter.Load(), decay, recordEntity.lastUpdated)
+		if err != nil {
+			return recordEntity, fmt.Errorf("could not decay cluster prefixed topic received counter: %w", err)
+		}
+		recordEntity.Counter.Store(decayedVal)
+		return recordEntity, nil
+	}
+}
+
+// activeClusterIdsKey returns the key used to store the active cluster ids in the cache.
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/cache.go]: Key was removed

File: network/p2p/inspector/internal/cache/active_cluster_ids_entity.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/active_cluster_ids_entity.go]: removes usage of atomic float64 https://github.com/onflow/flow-go/pull/4365/commits/b75b925021174bcc01153f7751e448e575124532

File: network/p2p/inspector/internal/cache/cache_test.go [hunk 13]
```diff
@@ -0,0 +1,549 @@
...
+		}()
+	}
+
+	unittest.RequireReturnsBefore(t, wg.Wait, 100*time.Millisecond, "timed out waiting for goroutines to finish")
+
+	// ensure that only one goroutine successfully initialized the record
+	require.Equal(t, int32(1), successGauge.Load())
+
+	// ensure that the record is correctly initialized in the cache
+	guage, found, _ := cache.Get(nodeID)
```
Comment (jordanschalm) [inline network/p2p/inspector/internal/cache/cache_test.go:139]: should check err

File: network/p2p/inspector/internal/cache/cache_test.go [hunk 14]
```diff
@@ -0,0 +1,549 @@
...
+	require.NoError(t, err)
+	require.Equal(t, float64(1), guage)
+
+	// get will apply a slightl decay resulting
+	// in a gauge value less than guage which is 1 but greater than 0.9
+	currentGauge, ok, err := cache.Get(nodeID1)
+	require.NoError(t, err)
+	require.True(t, ok)
+	require.LessOrEqual(t, currentGauge, guage)
+	require.Greater(t, currentGauge, 0.9999999)
```
Comment (jordanschalm) [inline network/p2p/inspector/internal/cache/cache_test.go:172]: My understanding is the decay is time-based. I'm wondering if there is an opportunity for creating a flaky test here on a slower machine? Should we just check that it's greater than 0.9?

File: network/p2p/inspector/internal/cache/cache_test.go [hunk 15]
```diff
@@ -0,0 +1,549 @@
...
+
+	// test adjusting the spam record for a non-existing node ID
+	nodeID3 := unittest.IdentifierFixture()
+	guage2, err := cache.Update(nodeID3)
```
Comment (jordanschalm) [inline network/p2p/inspector/internal/cache/cache_test.go:176]:
```suggestion
	guage3, err := cache.Update(nodeID3)
```
Not required - but personally I find it easier to follow when the number of different variables is consistent: "gauge N is the gauge for node ID N"

File: network/p2p/inspector/internal/cache/cache_test.go [hunk 16]
```diff
@@ -0,0 +1,549 @@
...
+	require.Equal(t, 3, len(identities))
+
+	identityMap := make(map[flow.Identifier]struct{})
+	for _, id := range identities {
+		identityMap[id] = struct{}{}
+	}
+
+	require.Contains(t, identityMap, nodeID1)
+	require.Contains(t, identityMap, nodeID2)
+	require.Contains(t, identityMap, nodeID3)
```
Comment (jordanschalm) [inline network/p2p/inspector/internal/cache/cache_test.go:247]:
```suggestion
	require.ElementsMatch(t, identities, []flow.Identifier{nodeID1, nodeID2, nodeID3})
```

File: network/p2p/inspector/internal/cache/cache_test.go [hunk 17]
```diff
@@ -0,0 +1,549 @@
...
+	nodeID1 := unittest.IdentifierFixture()
+	nodeID2 := unittest.IdentifierFixture()
+	nodeID3 := unittest.IdentifierFixture()
+
+	require.True(t, cache.Init(nodeID1))
+	require.True(t, cache.Init(nodeID2))
+	require.True(t, cache.Init(nodeID3))
+
+	numOfIds := uint(3)
+	require.Equal(t, numOfIds, cache.Size(), fmt.Sprintf("expected size of the cache to be %d", numOfIds+1))
```
Comment (jordanschalm) [inline network/p2p/inspector/internal/cache/cache_test.go:272]: outdated message

File: network/p2p/inspector/internal/cache/cache_test.go [hunk 18]
```diff
@@ -0,0 +1,549 @@
...
+}
+
+// TestRecordCache_Init tests the Init method of the RecordCache.
+// It ensures that the method returns true when a new record is initialized
+// and false when an existing record is initialized.
+func TestRecordCache_Init(t *testing.T) {
+	cache := cacheFixture(t, 100, defaultDecay, zerolog.Nop(), metrics.NewNoopCollector())
+	require.NotNil(t, cache)
+	// expect cache to be initialized with a empty active cluster IDs list
+	require.Equalf(t, uint(0), cache.Size(), "cache size must be 1")
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply network/p2p/inspector/internal/cache/cache_test.go:38]: We could also move the initial size check into `cacheFixture` and save these lines in all test cases

File: network/p2p/inspector/internal/cache/cache_test.go [hunk 19]
```diff
@@ -0,0 +1,549 @@
...
+			defer wg.Done()
+			removed := cache.Remove(id)
+			require.True(t, removed)
+			require.NotContains(t, id, cache.NodeIDs())
+		}(nodeID)
+	}
+
+	unittest.RequireReturnsBefore(t, wg.Wait, 100*time.Millisecond, "timed out waiting for goroutines to finish")
+
+	// ensure cache only has default active cluster Ids stored
```
Comment (jordanschalm) [inline network/p2p/inspector/internal/cache/cache_test.go:317]: oudated

File: network/p2p/inspector/internal/cache/cache_test.go [hunk 20]
```diff
@@ -0,0 +1,549 @@
...
+			_, err := cache.Update(id)
+			require.NoError(t, err)
+		}(nodeID)
+
+		// get spam records concurrently
+		go func(id flow.Identifier) {
+			defer wg.Done()
+			record, found, _ := cache.Get(id)
+			require.True(t, found)
+			require.NotNil(t, record)
```
Comment (jordanschalm) [inline network/p2p/inspector/internal/cache/cache_test.go:353]: record cannot be nil

File: network/p2p/inspector/internal/cache/cache_test.go [hunk 21]
```diff
@@ -0,0 +1,549 @@
...
+			cache.Remove(id)
+			require.NotContains(t, id, cache.NodeIDs())
+		}(nodeID)
+	}
+
+	unittest.RequireReturnsBefore(t, wg.Wait, 100*time.Millisecond, "timed out waiting for goroutines to finish")
+
+	// ensure that the initialized records are correctly added to the cache
+	// and removed records are correctly removed from the cache
+	require.Equal(t, uint(nodeIDsToAdd.Len()), cache.Size())
```
Comment (jordanschalm) [inline network/p2p/inspector/internal/cache/cache_test.go:413]:
```suggestion
	require.ElementsMatch(t, nodeIDsToAdd), cache.NodeIDs())
```

File: network/p2p/inspector/internal/cache/cache_test.go [hunk 22]
```diff
@@ -0,0 +1,549 @@
...
+
+	// Adjust spam records concurrently
+	for _, nodeID := range nodeIDsToAdjust {
+		go func(id flow.Identifier) {
+			defer wg.Done()
+			_, _ = cache.Update(id)
+		}(nodeID)
+	}
+
+	unittest.RequireReturnsBefore(t, wg.Wait, 100*time.Millisecond, "timed out waiting for goroutines to finish")
```
Comment (jordanschalm) [inline network/p2p/inspector/internal/cache/cache_test.go:464]: This test might be missing an assertion about the cache state after all the concurrent updates complete. Ie. it should contain exactly `nodeIDsToAdjust + nodeIDsToAdd`

File: network/p2p/inspector/internal/cache/cache.go [hunk 23]
```diff
@@ -0,0 +1,229 @@
...
+// Update applies an adjustment that increments the number of cluster prefixed control messages received by a peer.
+// Returns number of cluster prefix control messages received after the adjustment. The record is initialized before
+// the adjustment func is applied that will increment the Gauge.
+// Args:
+// - nodeID: the node ID of the sender of the control message.
+// - adjustFunc: the function that adjusts the record.
+// Returns:
+//   - The cluster prefix control messages received gauge value after the adjustment.
+//   - exception only in cases of internal data inconsistency or bugs. No errors are expected.
+func (r *RecordCache) Update(nodeID flow.Identifier) (float64, error) {
```
Comment (jordanschalm) [inline network/p2p/inspector/internal/cache/cache.go:92]:
```suggestion
func (r *RecordCache) ReceivedSpam(nodeID flow.Identifier) (float64, error) {
```
This module is very use-case-specific -- I feel we could have a more descriptive name than `Update`, to reflect that it is incrementing a counter, that it corresponds to receiving one "spam" message,  etc.

**Review by jordanschalm — APPROVED** [review]
My remaining comments are largely around updating the documentation and some minor suggestions. I personally think considering a more descriptive name than `Update` for noting a received spam message would be worthwhile, but I'll leave that for you and Yahya to decide. 

Great work on the well-documented and exhaustive cache tests  👏

File: network/p2p/inspector/internal/cache/cache.go [hunk 24]
```diff
@@ -48,163 +46,129 @@ type RecordCache struct {
...
 	// it means the record was not initialized. In this case, initialize the record and call optimisticAdjustFunc again.
 	// If the record was initialized, optimisticAdjustFunc will be called only once.
-	adjustedEntity, ok := optimisticAdjustFunc()
-	if !ok {
-		r.Init(originId)
-		adjustedEntity, ok = optimisticAdjustFunc()
-		if !ok {
-			return 0, fmt.Errorf("record not found for origin id %s, even after an init attempt", originId)
+	adjustedEntity, adjusted := optimisticAdjustFunc()
+	switch {
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/cache.go]: https://github.com/onflow/flow-go/pull/4365/commits/2aa17066fd892388e8b33aeb53899316978479be

File: network/p2p/inspector/internal/cache/tracker_test.go [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+			defer wg.Done()
+			_, err := tracker.Inc(id)
+			require.NoError(t, err)
+		}()
+	}
+	unittest.RequireReturnsBefore(t, wg.Wait, 100*time.Millisecond, "timed out waiting for goroutines to finish")
+	// after each decay is applied the gauge value result should be slightly less than n
+	gaugeVal, err := tracker.Load(id)
+	require.NoError(t, err)
+	require.True(t, n-gaugeVal < .2)
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/tracker_test.go]: https://github.com/onflow/flow-go/pull/4365/commits/45ad742b2800a250a22247c81c9c1e13e1782fc2

File: network/p2p/inspector/internal/cache/cache_test.go [hunk 26]
```diff
@@ -0,0 +1,549 @@
...
+	// expect cache to be initialized with a empty active cluster IDs list
+	require.Equalf(t, uint(0), cache.Size(), "cache size must be 1")
+}
+
+// TestRecordCache_Init tests the Init method of the RecordCache.
+// It ensures that the method returns true when a new record is initialized
+// and false when an existing record is initialized.
+func TestRecordCache_Init(t *testing.T) {
+	cache := cacheFixture(t, 100, defaultDecay, zerolog.Nop(), metrics.NewNoopCollector())
+	require.NotNil(t, cache)
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/cache_test.go]: https://github.com/onflow/flow-go/pull/4365/commits/911c471c12e9301d6ad9ae060cbca247d7136081

File: network/p2p/inspector/internal/cache/cache_test.go [hunk 27]
```diff
@@ -0,0 +1,549 @@
...
+			defer wg.Done()
+			cache.Init(id)
+		}(nodeID)
+	}
+
+	unittest.RequireReturnsBefore(t, wg.Wait, 100*time.Millisecond, "timed out waiting for goroutines to finish")
+
+	// ensure that all records are correctly initialized
+	for _, nodeID := range nodeIDs {
+		guage, found, _ := cache.Get(nodeID)
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/inspector/internal/cache/cache_test.go]: https://github.com/onflow/flow-go/pull/4365/commits/782fcee2181de9ee3c534cbce9b768ac9cf7901c

File: network/p2p/inspector/internal/cache/cache_test.go (hunk 15 above)
Comment (kc1116) [reply network/p2p/inspector/internal/cache/cache_test.go]: https://github.com/onflow/flow-go/pull/4365/commits/c523358691b76a89c49541cc90fd74a76437b867

File: network/p2p/inspector/internal/cache/cache_test.go (hunk 14 above)
Comment (kc1116) [reply network/p2p/inspector/internal/cache/cache_test.go]: https://github.com/onflow/flow-go/pull/4365/commits/1e87a3fe0f415769b0882837de49da354cc5b9a2

File: network/p2p/inspector/internal/cache/cache.go (hunk 23 above)
Comment (kc1116) [reply network/p2p/inspector/internal/cache/cache.go]: https://github.com/onflow/flow-go/pull/4365/commits/ac998fe58e12c603b56b53771ea83f2ce512d999

**Pull request merged**